warehouse/slave.go

```diff
@@ -189,126 +207,170 @@ func (job *Payload) getDiscardsTable() string {
-	return strings.TrimSuffix(jobRun.stagingFilePath, "json.gz") + tableName + fmt.Sprintf(`.%s`, randomness) + fmt.Sprintf(`.%s`, warehouseutils.GetLoadFileFormat(job.LoadFileType))
+	return strings.TrimSuffix(jr.stagingFilePath, "json.gz") + tableName + fmt.Sprintf(`.%s`, randomness) + fmt.Sprintf(`.%s`, warehouseutils.GetLoadFileFormat(job.LoadFileType))
 }
 
 func (job *Payload) getColumnName(columnName string) string {
 	return warehouseutils.ToProviderCase(job.DestinationType, columnName)
 }
 
-type loadFileUploadJob struct {
-	tableName  string
-	outputFile encoding.LoadFileWriter
-}
-
-type loadFileUploadOutput struct {
-	TableName             string
-	Location              string
-	TotalRows             int
-	ContentLength         int64
-	StagingFileID         int64
-	DestinationRevisionID string
-	UseRudderStorage      bool
-}
+// uploadLoadFiles returns the upload output for each file uploaded to object storage
+func (jr *JobRun) uploadLoadFiles(ctx context.Context) ([]loadFileUploadOutput, error) {
+	ctx, cancel := context.WithTimeout(ctx, jr.slaveUploadTimeout)
+	defer cancel()
 
-func (jobRun *JobRun) uploadLoadFilesToObjectStorage(ctx context.Context) ([]loadFileUploadOutput, error) {
-	job := jobRun.job
-	uploader, err := job.getFileManager(job.DestinationConfig, job.UseRudderStorage)
+	uploader, err := jr.job.getFileManager(jr.job.DestinationConfig, jr.job.UseRudderStorage)
 	if err != nil {
-		return []loadFileUploadOutput{}, err
-	}
-	// var loadFileIDs []int64
-	var loadFileUploadOutputs []loadFileUploadOutput
-
-	// take the first staging file id in upload
-	// TODO: support multiple staging files in one upload
-	stagingFileId := jobRun.job.StagingFileID
-
-	loadFileOutputChan := make(chan loadFileUploadOutput, len(jobRun.outputFileWritersMap))
-	loadFileUploadTimer := jobRun.timerStat("load_file_upload_time")
-	uploadJobChan := make(chan *loadFileUploadJob, len(jobRun.outputFileWritersMap))
-	// close chan to avoid memory leak ranging over it
-	defer close(uploadJobChan)
-	uploadErrorChan := make(chan error, numLoadFileUploadWorkers)
-	ctx, cancel := context.WithCancel(ctx)
-	defer cancel()
-	for i := 0; i < numLoadFileUploadWorkers; i++ {
-		go func(ctx context.Context) {
-			for uploadJob := range uploadJobChan {
+		return nil, fmt.Errorf("creating uploader: %w", err)
+	}
+
+	type result struct {
+		uploadOutput loadFileUploadOutput
+		err          error
+	}
+
+	type uploadJob struct {
+		tableName  string
+		outputFile encoding.LoadFileWriter
+	}
+
+	var output []loadFileUploadOutput
+	var totalUploadTime time.Duration
+	var mu sync.Mutex
+
+	defer func() {
+		if totalUploadTime > 0 {
+			jr.timerStat("load_file_upload_time").SendTiming(totalUploadTime)
+		}
+	}()
+
+	generate := func() <-chan *uploadJob {
+		jobStream := make(chan *uploadJob, len(jr.outputFileWritersMap))
+		go func() {
+			defer close(jobStream)
+
+			for tableName, loadFile := range jr.outputFileWritersMap {
 				select {
```

Comment:
With this pattern we are not honouring the context. We're doing:

```go
select {
case <-ctx.Done():
	return
default:
	jobStream <- &uploadJob{
		tableName:  tableName,
		outputFile: loadFile,
	}
}
```

We check `ctx.Done`, if not done we proceed with the `default`. In the `default` we try to push into the `jobStream` channel. The problem is that we cannot foresee how long the pushing is going to take. The context might get canceled before we have the chance to push into the channel. What's even worse is that, due to the same context cancelation, what if the reader of `jobStream` stops reading while the channel happened to be full? We would have a leak here since we would be stuck in the `default` unable to proceed with the execution in the routine.

In order to avoid the leak, all channel operations must be done via the `select`:

```go
select {
case <-ctx.Done():
	return
case jobStream <- &uploadJob{
	tableName:  tableName,
	outputFile: loadFile,
}:
}
```

It is true that any of the two cases of the `select` can be picked but that's unavoidable and not a problem in my opinion. This means that if both the context is done and `jobStream` is not closed (and has buffer available), we might push into the channel instead of returning due to the context cancelation. There is no way to prevent that but it's OK because A) we would still return and not block and B) right after pushing at the next iteration (or if there are no more iterations) we close `jobStream`.